### ISSUE TITLE
PT review: two-layer PTM is genuine, but the pseudo-doc assignment drops the (m_p + lambda) popularity prior (ships uniform-psi variant) and log_likelihood uses digamma not lgamma

### ISSUE THREAD

**Issue by nealcaren** [nealcaren/topica#491]
## Review result

Two independent reviewers (Claude subagent + Gemini/Antigravity — Codex stalled and was replaced) plus a maintainer trace, all agreeing. The headline good news: **topica's PT genuinely implements the two-layer PTM** — it is **not** a degenerate LDA-with-P=D. There is a real doc→pseudo-doc sampler (`resample_pseudo`) that migrates a whole document's topic counts between pseudo-documents, and a word→topic sampler (`resample_token`) that conditions on **per-pseudo-document** topic counts `n_{p,k}` (`src/pt.rs:167`), not per-doc counts. Decrement-before-resample, whole-document count migration, φ/θ derivation, hyperparameter guards, determinism, and the held-out `transform` all trace cleanly. Two fidelity items, one structural.

### 1. Pseudo-doc assignment drops the `(m_p + λ)` popularity prior — this is a uniform-ψ variant presented as the base PTM (MEDIUM-HIGH)

`src/pt.rs:191-236` (assignment log-prob at `222`). Zuo et al.'s PTM draws `l_d ~ Mult(ψ)` with `ψ ~ Dir(λ)` over the P pseudo-documents. Collapsing ψ makes the assignment conditional

```
P(l_d = p | ·) ∝ (m_p^{¬d} + λ) · Π_k Γ(n_{p,k}^{¬d}+m_{d,k}+α)/Γ(n_{p,k}^{¬d}+α) · Γ(n_p^{¬d}+Kα)/Γ(n_p^{¬d}+N_d+Kα)
```

where `m_p^{¬d}` is the number of **documents** currently assigned to pseudo-doc p. topica computes only the two Gamma-ratio content-fit factors and drops `(m_p^{¬d} + λ)`:

```rust
log_probs[p] = numer_log - denom_log;   // no + ln(m_p + λ) term
```

The omission is structural, and confirmed three ways:
- **No document-count-per-pseudo-doc array exists.** `np` holds pseudo-doc *token* totals (`self.np[p_old] -= doc.len()`, `pt.rs:206`), `npk` holds topic counts — there is no `m_p` (docs-per-pseudo-doc) array anywhere (`PtmModel` fields, `pt.rs:29-50`).
- **No `λ` / `pseudo_doc_prior` hyperparameter** on the constructor: `PT::new(num_topics, *, num_pseudo=100, alpha=0.1, beta=0.01, seed=42)` (`src/python/mod.rs:11196`).
- The internal Rust docstring already acknowledges it: "uniform prior over pseudo-docs, matching the simplest PTM variant — no pseudo-doc count prior term" (`pt.rs:189-190`).

The Gamma-ratio (Pochhammer) content terms themselves are correct in log space (`pt.rs:214-221`), with a max-shift softmax.

**Consequence.** The base PTM's rich-get-richer aggregation — popular pseudo-docs attract more documents, with λ preventing collapse — is entirely absent; assignment is driven purely by topic-content fit, so the stationary distribution and recovered pseudo-doc structure differ from a faithful PTM, most on corpora where the aggregation dynamics matter. The class docstring, the `.pyi` stub, and the model name present this as "Pseudo-document Topic Model (Zuo et al. 2016)" with no user-facing note of the simplification.

- [ ] Either **(a) faithful:** track `m_p` (docs per pseudo-doc), add a `pseudo_doc_prior`/`lambda` hyperparameter, and add `ln(m_p^{¬d} + λ)` to `log_probs[p]`, decrementing/incrementing `m_p` alongside the token migration; or **(b) honest-as-is:** relabel/redocument the class docstring, `.pyi`, and docs as the uniform-ψ ("SPTM-style") variant so the "Zuo et al. 2016 PTM" claim isn't overstated (mirrors the parity-wording fixes in #490/#489). The planted self-consistency gold (`parity/pt_gold.py`) cannot surface this — no external PTM to compare against.

### 2. `log_likelihood()` uses `digamma` where a collapsed log-likelihood needs `lgamma` (LOW-MED)

`src/pt.rs:264-301`. The function is documented as a "Collapsed Gibbs log-likelihood" and feeds `fit_history` and the `convergence_tol` early-stop (`pt.rs:435-443`), but every term uses `digamma` (ψ, the *derivative* of lnΓ) rather than `lgamma`:

```rust
ll += digamma(self.alpha + n_pk) - digamma(self.alpha);   // pt.rs:282
ll -= digamma(k_alpha + n_p)  - digamma(k_alpha);         // pt.rs:285
ll += digamma(self.beta + n_kw) - digamma(self.beta);     // pt.rs:293
ll -= digamma(v_beta + self.nk[kk] as f64) - digamma(v_beta); // pt.rs:296
```

A collapsed Dirichlet-multinomial marginal is a sum of `lnΓ` terms; `digamma` appears in gradients/hyperparameter updates, not in a joint log-probability. digamma is increasing, so the quantity trends upward as the model concentrates and works as a crude monotone diagnostic, but it is not the log marginal likelihood, and `convergence_tol` thresholds relative change of this surrogate — so the early-stop trips on a non-LL quantity. The docstring hedges ("a cheap and informative per-sweep objective; it does not integrate the pseudo-doc assignment"), but the `log_likelihood` name and "log-likelihood" prose overstate it.

- [ ] Switch the terms to `lgamma` (a genuine collapsed LL up to constants), or rename to `objective`/`fit_score` and drop the "log-likelihood" wording.

### 3. Minor / informational

- **`num_pseudo=100` default can exceed the corpus size** (`src/python/mod.rs:11196`). The paper's regime is P ≪ D; for D < 100 the default gives P ≥ D, so most pseudo-docs hold ≤1 real doc and the aggregation that is PTM's whole purpose collapses toward per-document LDA, with no warning. Consider warning (not erroring) when `num_pseudo >= num_docs` at fit time, or documenting that P should be well below D.
- **`transform` bypasses the pseudo-doc layer** (`src/python/mod.rs:11475-11520`) — held-out docs infer θ over K topics directly against fixed φ, with no pseudo-doc assignment. Correct and honestly documented (the paper does not define held-out pseudo-doc assignment); noted for completeness only.

### Checked and CORRECT

Two genuine latent layers (`resample_token` conditions on `npk[p][kk]`, `pt.rs:167`; `resample_pseudo` reassigns on the joint likelihood of the whole doc's topics); decrement-before-resample for tokens and whole-document count migration for pseudo-docs; the Gamma-ratio assignment term is the correct Pochhammer form in stable log space; θ semantics (each real doc inherits its pseudo-doc's `(n_{p,k}+α)/(n_p+Kα)` — docs in a pseudo-doc share a θ row, expected); φ = `(n_{kw}+β)/(n_k+Vβ)` sums to 1; defaults α=0.1, β=0.01 consistent with the short-text convention; hyperparameter guards route α/β through `finite_pos` (#481 class), `num_topics≥2`, `num_pseudo≥1`; robustness (empty docs → uniform pick no NaN, P=1, empty pseudo-doc stays selectable, no div-by-zero); determinism via `Pcg64Mcg::seed_from_u64`; `theta_draws` ring-buffer; save/load persists the model with `PtState`; stub in sync.

### Bottom line

The pseudo-document layer is real and correctly wired — a faithful *two-level* model, not LDA in disguise. The single meaningful fidelity gap is **#1**: the assignment sampler ships a fixed-uniform-ψ while the model presents itself as Zuo et al.'s PTM. Highest-value action is to restore the popularity term (with `λ` and an `m_p` count) or, at minimum, correct the user-facing docstring/stub/name — plus relabel the `digamma` `log_likelihood` (#2).

Relevant surfaces: `src/pt.rs`, `src/python/mod.rs:11131-11530`, `python/topica/_topica.pyi`, `parity/pt_gold.py`, `tests/test_pt_gold.py`.

**Comment by nealcaren** [comment]
The two fidelity items landed in #511 (merged): (1) the faithful `(m_p + λ)` popularity prior — PT now tracks docs-per-pseudo-doc `mp` and adds `ln(m_p + λ)` to the assignment log-prob, with a `pseudo_doc_prior` kwarg (default 0.1); (2) `log_likelihood` switched digamma→lgamma (a genuine collapsed DM marginal). Regime note: the faithful prior is init-sensitive at scarce P (P=10 recovers ~4/8 seeds; P=50 → 8/8), so the gold/recovery-test moved to P=50. Still open: the `num_pseudo >= num_docs` warning (now more salient given the P-sensitivity) and the transform-bypasses-pseudo-doc note.
